spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala

```diff
@@ -0,0 +1,376 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.stats
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import org.apache.spark.sql.delta.actions.{Action, AddFile, FileAction, RemoveFile}
+import org.apache.spark.sql.delta.hooks.AutoCompactPartitionReserve
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+import org.apache.spark.sql.SparkSession
+
+/**
+ * A collector used to aggregate auto-compaction stats for a single commit. The expectation
+ * is to spin this up for a commit and then merging those local stats with the global stats.
+ */
+trait AutoCompactPartitionStatsCollector {
+  def collectPartitionStatsForAdd(file: AddFile): Unit
+  def collectPartitionStatsForRemove(file: RemoveFile): Unit
+  def finalizeStats(tableId: String): Unit
+}
+
+/**
+ * This singleton object collect the table partition statistic for each commit that creates
+ * AddFile or RemoveFile objects.
+ * To control the memory usage, there are `maxNumTablePartitions` per table and 'maxNumPartitions'
+ * partition entries across all tables.
+ * Note:
+ *   1. Since the partition of each table is limited, if this limitation is reached, the least
+ *      recently used table partitions will be evicted.
+ *   2. If all 'maxNumPartitions' are occupied, the partition stats of least recently used tables
+ *      will be evicted until the used partitions fall back below to 'maxNumPartitions'.
+ *   3. The un-partitioned tables are treated as tables with single partition.
+ * @param maxNumTablePartitions The hash space of partition key to reduce memory usage per table.
+ * @param maxNumPartitions The maximum number of partition that can be occupied.
+ */
+class AutoCompactPartitionStats(
+    private var maxNumTablePartitions: Int,
+    private var maxNumPartitions: Int
+) {
+
+  /**
+   * This class to store the states of one table partition. These state includes:
+   * -- the number of small files,
+   * -- the thread that assigned to compact this partition, and
+   * -- whether the partition was compacted.
+   *
+   * Note: Since this class keeps tracking of the statistics of the table partition and the state of
+   * the auto compaction thread that works on the table partition, any method that accesses any
+   * attribute of this class needs to be protected by synchronized context.
+   */
+  class PartitionStat(
+      var numFiles: Long,
+      var wasAutoCompacted: Boolean = false) {
+
+    /**
+     * Determine whether this partition can be autocompacted based on the number of small files or
+     * if this [[AutoCompactPartitionStats]] instance has not auto compacted it yet.
+     * @param minNumFiles The minimum number of files this table-partition should have to trigger
+     *                    Auto Compaction in case it has already been compacted once.
+     */
+    def hasSufficientSmallFilesOrHasNotBeenCompacted(minNumFiles: Long): Boolean =
```

Comment:
`...OrNotCompactedYet`?

Reply:
re https://github.com/delta-io/delta/pull/2414#issuecomment-1883922342, perhaps we can address post merge